File: helper/src/main/java/com/usage/access/helper/UsageAccessHelper.java
package com.usage.access.helper;


import android.app.AppOpsManager;
import android.content.Context;
import android.content.Intent;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Build;
import android.os.Handler;
import android.os.Looper;
import android.provider.Settings;
import android.util.Log;
import android.widget.Toast;

public class UsageAccessHelper {
    private static final String TAG = "usage.access.helper";
    private final Context activity;
    private final Handler handler = new Handler(Looper.getMainLooper());

    public UsageAccessHelper(Context activity) {
        this.activity = activity;
    }

    private void makeToast(String m) {
        Toast.makeText(activity, m, Toast.LENGTH_SHORT).show();
    }

    private void log(String m) {
        Log.v(TAG, m);
    }

    public boolean getAccessGranted() {
        boolean result = false;
        try {
            PackageManager packageManager = activity.getPackageManager();
            ApplicationInfo applicationInfo = packageManager.getApplicationInfo(activity.getPackageName(), 0);
            AppOpsManager appOpsManager = (AppOpsManager) activity.getSystemService(Context.APP_OPS_SERVICE);
            int mode = appOpsManager.checkOpNoThrow(AppOpsManager.OPSTR_GET_USAGE_STATS, applicationInfo.uid, applicationInfo.packageName);
            result = mode == 0;
        } catch (PackageManager.NameNotFoundException ignored) { }
        return result;
    }

    public final boolean permitUsageAccess(boolean forceToShow) {
        boolean access = getAccessGranted();
        if (access && !forceToShow)
            return true;
        handler.post(() -> {
            waitForUsageAccessPermissionGranted();
            Intent intent = new Intent(Settings.ACTION_USAGE_ACCESS_SETTINGS);
            if (Build.VERSION.SDK_INT > Build.VERSION_CODES.P) {
                intent.setData(Uri.parse("package:" + activity.getPackageName()));
            }
            log("pre-start settings activity");
            activity.startActivity(intent);
            makeToast("Settings activity started!");
        });

        return access;
    }

    private void waitForUsageAccessPermissionGranted() {
        AppOpsManager appOpsManager = (AppOpsManager) activity.getSystemService(Context.APP_OPS_SERVICE);
        AppOpsManager.OnOpChangedListener watcher = new AppOpsManager.OnOpChangedListener() {
            @Override
            public void onOpChanged(String s, String s1) {
                if (!getAccessGranted())
                    return;
                handler.post(() -> {
                    log("Try to launch activity!");
                    appOpsManager.stopWatchingMode(this);
                    Intent intent = new Intent(activity, activity.getClass());
                    intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_SINGLE_TOP);
                    activity.startActivity(intent);
                    log("Unity Activity launched success!");
                });
            }
        };
        appOpsManager.startWatchingMode(
                AppOpsManager.OPSTR_GET_USAGE_STATS,
                activity.getPackageName(),
                watcher
        );
    }

}

File: usage-access-permission-helper/src/main/java/usage/access/permission/helper/UsageAccessHelper.kt
package usage.access.permission.helper

import android.app.Activity
import android.app.AppOpsManager
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Build
import android.provider.Settings
import android.util.Log
import android.widget.Toast
import androidx.activity.ComponentActivity
import androidx.core.content.ContextCompat.startActivity
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch

class UsageAccessHelper(
    /**
     * Unity main activity
     */
    private val activity: Context
) {
    val accessGranted: Boolean
        get() {
            return try {
                val packageManager = activity.packageManager
                val applicationInfo = packageManager.getApplicationInfo(activity.packageName, 0)
                val appOpsManager =
                    activity.getSystemService(ComponentActivity.APP_OPS_SERVICE) as AppOpsManager
                val mode = appOpsManager.checkOpNoThrow(
                    AppOpsManager.OPSTR_GET_USAGE_STATS,
                    applicationInfo.uid, applicationInfo.packageName
                )
                mode == AppOpsManager.MODE_ALLOWED
            } catch (e: PackageManager.NameNotFoundException) {
                false
            }
        }

    /**
     * Prompt usage access permission from user.
     * @forceToShow show settings activity always (even if permission is granted)
     * @return has permission already granted?
     */
    @JvmOverloads
    fun permitUsageAccess(forceToShow: Boolean = false): Boolean {
        val access = accessGranted
        if (access && !forceToShow)
            return true

        waitForUsageAccessPermissionGranted()
        val intent = Intent(Settings.ACTION_USAGE_ACCESS_SETTINGS)
        if (Build.VERSION.SDK_INT > Build.VERSION_CODES.P)
            intent.data = Uri.parse("package:${activity.packageName}")
//        intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK // Чтобы не оставалось в "последних" при возврате в приложение
        activity.startActivity(intent)
        return access
    }

    private fun waitForUsageAccessPermissionGranted(): Unit {
        val appOpsManager =
            activity.getSystemService(ComponentActivity.APP_OPS_SERVICE) as AppOpsManager

        fun grantPermissionWatcher(op: String, packageName: String) {
            if (!accessGranted)
                return

            GlobalScope.launch(Dispatchers.Main) {
                log("Try to launch activity!")
                appOpsManager.stopWatchingMode(::grantPermissionWatcher)
                val intent = Intent(activity, activity.javaClass)
                intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
                activity.startActivity(intent)
                log("Activity launched success!")
            }
        }

        appOpsManager.startWatchingMode(
            AppOpsManager.OPSTR_GET_USAGE_STATS,
            activity.packageName,
            ::grantPermissionWatcher
        )
    }
}